iceoryx_hoofs/test/moduletests/test_cxx_function.cpp

```diff
@@ -0,0 +1,630 @@
+using fixed_size_function = iox::cxx::function<T, Bytes>;
+using test_function = fixed_size_function<signature>;
+
+
+// helper template to count construction and copy statistics,
+// for our purpose (test_function) we do not need to distinguish
+// between copy (move) ctor and copy(move) assignment
+template <typename T>
+class Counter
+{
+  public:
+    static uint64_t numCreated;
+    static uint64_t numCopied;
+    static uint64_t numMoved;
+    static uint64_t numDestroyed;
+
+    Counter()
+    {
+        ++numCreated;
+    }
+
+    Counter(const Counter&)
+    {
+        ++numCreated;
+        ++numCopied;
+    }
+
+    Counter(Counter&&)
+    {
+        ++numMoved;
+    }
+
+    ~Counter()
+    {
+        ++numDestroyed;
+    }
+
+    Counter& operator=(const Counter&)
+    {
+        ++numCopied;
+    }
+
+    Counter& operator=(Counter&&)
+    {
+        ++numMoved;
+    }
+
+    static void resetCounts()
+    {
+        numCreated = 0U;
+        numCopied = 0U;
+        numMoved = 0U;
+        numDestroyed = 0U;
+    }
+};
+
+template <typename T>
+uint64_t Counter<T>::numCreated = 0U;
+
+template <typename T>
+uint64_t Counter<T>::numCopied = 0U;
+
+template <typename T>
+uint64_t Counter<T>::numMoved = 0U;
+
+template <typename T>
+uint64_t Counter<T>::numDestroyed = 0U;
+
+
+class Functor : public Counter<Functor>
+{
+  public:
+    Functor(int32_t state)
+        : Counter<Functor>()
+        , m_state(state)
+    {
+    }
+
+    int32_t operator()(int32_t n)
+    {
+        m_state += n;
+        return m_state;
+    }
+
+    // integer arg to satisfy signature requirement of our test_function
+    int32_t getState(int32_t n = 0) const
+    {
+        return m_state + n;
+    }
+
+    int32_t m_state{0};
+};
+
+int32_t freeFunction(int32_t n)
+{
+    return n + 1;
+};
+
+struct Arg : Counter<Arg>
+{
+    Arg() = default;
+    Arg(uint32_t value)
+        : value(value){};
+    Arg(const Arg&) = default;
+    Arg& operator=(const Arg&) = default;
+
+    // We cannot delete the move ctor, the function wrapper requires the arguments to be copy-constructible.
+    // According to the standard this means the copy Ctor must exist and move cannot be explicitly deleted.
+    // Move does not necessarily have to be defined, in which case the compiler will perform a copy
+    // whenever a move would be possible.
+    // Note that this is mainly an issue if the argument is passed by value.
+    // The std::function also fails to compile in this case (gcc implementation).
+
+    int32_t value;
+};
+
+int32_t freeFunctionWithCopyableArg(Arg arg)
+{
+    return arg.value;
+}
+
+
+class function_test : public Test
+{
+  public:
+    void SetUp() override
+    {
+    }
+
+    void TearDown() override
+    {
+    }
+
+    static int32_t staticFunction(int32_t n)
+    {
+        return n + 1;
+    }
+};
+
+TEST_F(function_test, DefaultConstructionCreatesNoCallable)
+{
+    test_function sut;
+
+    EXPECT_FALSE(sut.operator bool());
+}
+
```

Comment:
You could add a death that verifies the termination behaviour in case an empty `function` is invoked.

Reply:
No, should we decide to have Expects not terminate the test will be invalid. I would not make this terminate a requirement with such a test.